### PR TITLE
first draft of bringing workspace URLs into the Omnibox

### PR DESCRIPTION
**Pull request by pvh**
Let's get this landed -- we have in here a "WorkspaceInList" widget, a bigger omnibox, and support for having multiple workspaces to switch between.

Not included here:
 - the actual workspace design for omnibox
 - workspace creation (you gotta go get one)
 - organization names for contacts (to help differentiate your workspaces)
 - workspace deletion
 - [known issue] arrow keys don't scroll the workspace to the new highlighted item

### CODE REVIEW

File: src/renderer/components/Root.tsx [hunk 1]
```diff
@@ -26,27 +27,34 @@ import './content-types/files/AudioContent'
 import './content-types/files/VideoContent'
 import './content-types/files/PdfContent'
 import System, { SystemContext } from '../System'
+<<<<<<< HEAD
```
Comment (mjtognetti) [inline src/renderer/components/Root.tsx]: heads up, conflict markers

File: src/renderer/components/content-types/defaults/DefaultInList.tsx [hunk 2]
```diff
@@ -27,7 +27,7 @@ export default function ListItem(props: ContentProps) {
 
   // TODO: pick background color based on url
   return (
-    <div className="DocLink" style={css.listItem}>
```
Comment (pvh) [inline src/renderer/components/content-types/defaults/DefaultInList.tsx]: totally unrelated, but i noticed on the workspaceinlist that this same style didn't match other items and it was an uncool CSS object anyway, so i cut it.

File: src/renderer/components/content-types/workspace/Omnibox.css [hunk 3]
```diff
@@ -38,6 +38,12 @@
   flex: 1;
 }
 
+.Omnibox .ListMenu {
+  color: var(--colorBlueBlack);
+  height: 60vh;
+  overflow-y: scroll;
```
Comment (pvh) [inline src/renderer/components/content-types/workspace/Omnibox.css]: we probably want nicer scrollbars. i'll add that to a follow-up TODO list

File: src/renderer/components/content-types/workspace/Omnibox.tsx [hunk 4]
```diff
@@ -432,6 +435,15 @@ export default class Omnibox extends React.PureComponent<Props, State> {
           .filter(([id, doc]) => doc.name.match(new RegExp(state.search, 'i')))
           .map(([id, doc]) => ({ url: createDocumentLink('contact', id as HypermergeUrl) })),
     },
+    {
```
Comment (pvh) [inline src/renderer/components/content-types/workspace/Omnibox.tsx]: this is a temporary home, but works fine for now.

File: src/renderer/components/content-types/workspace/Workspace.tsx [hunk 5]
```diff
@@ -22,6 +22,8 @@ import { BoardDoc, CardId } from '../board'
 import { useSystem } from '../../../System'
 import { CurrentDeviceContext } from './Device'
 
+import WorkspaceInList from './WorkspaceInList'
```
Comment (pvh) [inline src/renderer/components/content-types/workspace/Workspace.tsx]: at some point soon i'll move this to `workspace/index.ts` as per board and contact

File: src/renderer/components/content-types/workspace/WorkspaceInList.css [hunk 6]
```diff
@@ -0,0 +1,15 @@
+.WorkspaceLink-Badge {
+  position: relative;
+}
+.WorkspaceLink-ContactOverlay {
+  position: absolute;
+  right: 0px;
+  bottom: 0px;
+  transform: translate(8px, 8px) scale(0.5);
```
Comment (pvh) [inline src/renderer/components/content-types/workspace/WorkspaceInList.css]: i'm not crazy about this approach but maybe it's okay? opinions invited.

Comment (pvh) [inline src/renderer/components/content-types/workspace/WorkspaceInList.css]: i'm not crazy about this approach but maybe it's okay? opinions invited.

File: src/renderer/components/content-types/workspace/WorkspaceInList.tsx [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        <Badge icon="briefcase" backgroundColor={selfDoc.color} />
+        <div className="WorkspaceLink-ContactOverlay">
+          <Content url={createDocumentLink('contact', selfId)} context="title-bar" />
+        </div>
+      </div>
+      <div className="DocLink__title">
+        <div className="WorkspaceLink-words">
+          <Text>Workspace for {selfDoc.name}</Text>
+          <SecondaryText>
+            {viewedDocUrls.length} item{viewedDocUrls.length !== 1 ? 's' : ''}
```
Comment (pvh) [inline src/renderer/components/content-types/workspace/WorkspaceInList.tsx]: this is bad i18n but.. meh?

File: src/renderer/WorkspaceHooks.ts [hunk 8]
```diff
@@ -1,46 +1,74 @@
...
-      saveWorkspaceUrl(workspaceUrl)
+    if (workspaceUrls.length > 0) {
+      saveWorkspaceUrls(workspaceUrls)
     }
-  }, [workspaceUrl])
+  }, [workspaceUrls])
+
+  const addWorkspaceUrl = useCallback(
+    (workspaceUrl) => {
+      const newWorkspaceUrls = [workspaceUrl, ...workspaceUrls.filter((u) => u !== workspaceUrl)]
```
Comment (mjtognetti) [inline src/renderer/WorkspaceHooks.ts]: Is there any value to the ordering of the workspace urls?

File: src/renderer/components/content-types/workspace/Omnibox.tsx [hunk 9]
```diff
@@ -432,6 +435,16 @@ export default class Omnibox extends React.PureComponent<Props, State> {
...
+      name: 'workspaces',
+      label: 'Workspaces',
+      actions: [this.view],
+      items: (state, props) =>
+        // we hide workspace results during a search right now
+        !this.props.workspaceUrlsContext || state.search
```
Comment (mjtognetti) [inline src/renderer/components/content-types/workspace/Omnibox.tsx]: some parens would go a long way towards readability here!

File: src/renderer/WorkspaceHooks.ts (hunk 8 above)
Comment (pvh) [reply src/renderer/WorkspaceHooks.ts]: well, aside from the current workspace being the most recently set one, it seemed reasonable to me to sort them by the order they were last used.

File: src/renderer/components/content-types/workspace/Omnibox.tsx (hunk 9 above)
Comment (pvh) [reply src/renderer/components/content-types/workspace/Omnibox.tsx]: i agree... but computer says no
(prettier takes them off.)

**Comment by mjtognetti** [comment]
Worth noting that we show avatar/icon at the top of the omnibox:
Screen Shot 2019-10-02 at 3 15 07 PM

and in the omnibox search we show the icon/avatar:
Screen Shot 2019-10-02 at 3 14 23 PM

Confusing to show both.

**Comment by mjtognetti** [comment]
Also worth noting, there is a placement issue with the avatar in the icon/avatar badge (note the white pixels in the lower righthand side of the avatar in the photo above). Don't believe this patch introduced it, but calling out.

File: src/renderer/components/content-types/workspace/WorkspaceInList.tsx [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+export default function WorkspaceListItem(props: ContentProps) {
+  const [doc] = useDocument<Doc>(props.hypermergeUrl)
+  const [selfDoc] = useDocument<ContactDoc>(doc && doc.selfId)
+
+  if (!doc || !selfDoc) {
+    return null
+  }
+
+  const { selfId, viewedDocUrls } = doc
```
Comment (mjtognetti) [inline src/renderer/components/content-types/workspace/WorkspaceInList.tsx]: wonky wonky hooks

**Pull request merged**